tiledb/sm/subarray/subarray.cc

```diff
@@ -783,6 +836,113 @@ const T* Subarray::tile_coords_ptr(
 }
 
+Status Subarray::compute_relevant_fragment_est_result_sizes(
+    const std::vector<std::string>& names,
+    uint64_t range_start,
+    uint64_t range_end,
+    std::vector<std::vector<ResultSize>>* result_sizes,
+    std::vector<std::vector<MemorySize>>* mem_sizes) {
+  // For easy reference
+  auto array_schema = array_->array_schema();
+  auto fragment_metadata = array_->fragment_metadata();
+  auto encryption_key = array_->encryption_key();
+  auto dim_num = array_->array_schema()->dim_num();
+  auto layout = (layout_ == Layout::UNORDERED) ? cell_order_ : layout_;
+
+  RETURN_NOT_OK(load_relevant_fragment_tile_var_sizes(names));
+
+  // Prepare result sizes vectors
+  auto range_num = range_end - range_start + 1;
+  result_sizes->resize(range_num);
+  std::vector<std::set<std::pair<unsigned, uint64_t>>> frag_tiles(range_num);
+  for (size_t r = 0; r < range_num; ++r)
+    (*result_sizes)[r].reserve(names.size());
+
+  // Create vector of var sizes
+  std::vector<bool> var_sizes;
+  var_sizes.reserve(names.size());
+  for (const auto& name : names)
+    var_sizes.push_back(array_schema->var_size(name));
+
+  auto all_dims_same_type = array_schema->domain()->all_dims_same_type();
+  auto all_dims_fixed = array_schema->domain()->all_dims_fixed();
+  auto num_threads = array_->num_threads();
+  auto ranges_per_thread = (uint64_t)ceil((double)range_num / num_threads);
+  auto statuses = parallel_for(0, num_threads, [&](uint64_t t) {
+    auto r_start = range_start + t * ranges_per_thread;
+    auto r_end =
+        std::min(range_start + (t + 1) * ranges_per_thread - 1, range_end);
+    auto r_coords = get_range_coords(r_start);
+    for (uint64_t r = r_start; r <= r_end; ++r) {
+      RETURN_NOT_OK(compute_relevant_fragment_est_result_sizes(
+          encryption_key,
+          array_schema,
+          all_dims_same_type,
+          all_dims_fixed,
+          fragment_metadata,
+          names,
+          var_sizes,
+          r,
+          r_coords,
+          &(*result_sizes)[r - range_start],
+          &frag_tiles[r - range_start]));
+
+      // Get next range coordinates
+      if (layout == Layout::ROW_MAJOR) {
+        auto d = dim_num - 1;
+        ++(r_coords)[d];
+        while ((r_coords)[d] >= ranges_[d].size() && d != 0) {
+          (r_coords)[d] = 0;
+          --d;
+          ++(r_coords)[d];
+        }
+      } else if (layout == Layout::COL_MAJOR) {
+        auto d = (unsigned)0;
+        ++(r_coords)[d];
+        while ((r_coords)[d] >= ranges_[d].size() && d != dim_num - 1) {
+          (r_coords)[d] = 0;
+          ++d;
+          ++(r_coords)[d];
+        }
+      } else {
+        // Global order - noop
+      }
+    }
+    return Status::Ok();
+  });
+  for (auto st : statuses)
+    RETURN_NOT_OK(st);
+
+  // Compute the mem sizes vector
+  mem_sizes->resize(range_num);
+  for (auto& ms : *mem_sizes)
+    ms.resize(names.size(), {0, 0});
+  std::set<std::pair<unsigned, uint64_t>> all_frag_tiles;
```

Comment:
since we don't need ordering, prefer `unordered_set` here

Reply:
Done.